README.md
| Travis        | AppVeyor      | GitLab| Codecov|
|:-------------:|:-------------:|:-----:|:------:|
| [![Build Status](https://travis-ci.org/Exiv2/exiv2.svg?branch=master)](https://travis-ci.org/Exiv2/exiv2) | [![Build status](https://ci.appveyor.com/api/projects/status/d6vxf2n0cp3v88al/branch/master?svg=true)](https://ci.appveyor.com/project/piponazo/exiv2-wutfp/branch/master) | [![pipeline status](https://gitlab.com/D4N/exiv2/badges/master/pipeline.svg)](https://gitlab.com/D4N/exiv2/commits/master) | [![codecov](https://codecov.io/gh/Exiv2/exiv2/branch/master/graph/badge.svg)](https://codecov.io/gh/Exiv2/exiv2) |

![Exiv2](http://www.exiv2.org/include/exiv2-logo-big.png)

Welcome to Exiv2, a C++ library and a command line utility to read and
write Exif, IPTC and XMP image metadata. The homepage of Exiv2 is: https://github.com/Exiv2/exiv2

## Source Code

Download the source code [Source](http://exiv2.dyndns.org/0.27.0.1/bundles/exiv2-0.27.0.1-Source.tar.gz)


## Documentation
Exiv2 API and tag reference documentation is at [Docs](http://exiv2.dyndns.org/0.27.0.1/doc/index.html)

## Building and Installing

Please read [README-CMAKE](http://exiv2.dyndns.org/0.27.0.1/README-CMAKE.md) for more information about how to build Exiv2.

### To install Exiv2 from a UNIX-like system, run:

    $ cd build
    $ sudo make install

### The default install locations are:

| Location       | Purpose           |
|:-------------  |:------------- |
| /usr/local/lib | for the library |
| /usr/local/bin | for the `exiv2` utility |
| /usr/local/include/exiv2 | for header files |

### To uninstall Exiv2 from a UNIX-like system, run:

    $ sudo make uninstall

## Dependencies

You can choose to use dependent libraries using your platform's package installer.

We also support conan which is especially useful for users of Visual Studio.
See [README-CONAN](http://exiv2.dyndns.org/0.27.0.1/README-CONAN.md) for more information

The following libexiv2 features are enabled by default and may
require external libraries. You can disable the dependency with CMake options

| Feature                    | Package  |  cmake option                | Availability |
|:-------------------------- |:-------- |:---------------------------- |:----------- |
|PNG image support           | zlib      | `-DEXIV2_ENABLE_PNG=Off`     | http://www.gnu.org/software/gettext/ |
|Native language support     | gettext   | `-DEXIV2_ENABLE_NLS=Off`     | http://zlib.net/ |
|XMP support                 | expat     | `-DEXIV2_ENABLE_XMP=Off`     | http://expat.sourceforge.net/ |

On Linux, it is usually best to install the dependencies through the
package management system of the distribution together with the
corresponding development packages (for the header files and static
libraries).

## Building Exiv2 Documentation

To generate the documentation (`make doc`), you will further need:
	`$ cmake ..options.. -DEXIV2_BUILD_DOC=On`

To build the documentation, you will need the following products:

| Product                    | Availability |
|:-------------------------- |:----------- |
| doxygen        |     http://www.doxygen.org/ |
|	graphviz     |    http://www.graphviz.org/ |
|	python       |   http://www.python.org/    |
|	xsltproc     |  http://xmlsoft.org/XSLT/   |
|	md5sum       | http://www.microbrew.org/tools/md5sha1sum/ |
|	pkg-config   | http://pkg-config.freedesktop.org/wiki/ |

## Support

All project resources are accessible from the project website.
    https://github.com/Exiv2/exiv2

For new bug reports and feature requests, please open an issue in Github.

## Download built libraries and linking your own code with Exiv2

You can download pre-built 'bundles' which include documentation.

[[Linux](http://exiv2.dyndns.org/0.27.0.1/bundles/exiv2-0.27.0.1-Linux.tar.gz)]
[[MacOSX](http://exiv2.dyndns.org/0.27.0.1/bundles/exiv2-0.27.0.1-Darwin.tar.gz)]
[[Cygwin](http://exiv2.dyndns.org/0.27.0.1/bundles/exiv2-0.27.0.1-CYGWIN.tar.gz)]
[[MinGW](http://exiv2.dyndns.org/0.27.0.1/bundles/exiv2-0.27.0.1-MinGW.tar.gz)]
[[MSVC](http://exiv2.dyndns.org/0.27.0.1/bundles/exiv2-0.27.0.1-msvc.zip)]

The file "ReadMe.txt" in a bundle describes how to install/link code with libraries for their respective platform.
The file "license.txt" in a bundle describes is a copy of GPLv2 License



## License

Copyright (C) 2004-2018 Exiv2 authors

Exiv2 is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your
option) any later version.

Exiv2 is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, 5th Floor, Boston,
MA 02110-1301 USA.

## Platform Notes
These are rough notes, to be polished.

### MinGW msys/2 64

https://www.msys2.org
Download:  http://repo.msys2.org/distrib/x86_64/msys2-x86_64-20180531.exe

#### msys64.bat

```
554 rmills@rmillsmm-w7:/c/Users/rmills/com $ cat msys64.bat
setlocal
set "PATH=c:\msys64\usr\bin;c:\msys64\usr\local\bin;c:\msys64\mingw64\bin;"
set "PS1=\! \u@\h:\w \$ "
bash.exe -norc
555 rmills@rmillsmm-w7:/c/Users/rmills/com $
```

#### Install MinGW Dependencies

```
$ for i in mingw-w64-x86_64-gcc mingw-w64-x86_64-gdb mingw-w64-x86_64-make mingw-w64-x86_64-pkg-config mingw-w64-x86_64-cmake; do (echo y|pacman -S $i); done
```

#### Download exiv2 from github

```
$ mkdir -p ~/gnu/github/exiv2
$ git clone https://github.com/exiv2/exiv2
$ cd exiv2
$ mkdir build ; cd build ;
$ cmake .. -G "Unix Makefiles"
$ make
```

### Cygwin

Download: https://cygwin.com/install.html and run setup-x86_64.exe

You need:
cmake, gcc, pkg-config, dos2unix, zlib-devel, libexpat-devel, make, git, python3-interpreter, libiconv, libxml2-utils

To build cmake (from 3.11.2), you need libncurses

### Linux


```
sudo apt-get update
sudo apt-get install build-essential

gcc,  	libxml2-utils
```


releasenotes/releasenotes.txt
Exiv2 v0.27 RC1 (exiv2 v0.27.0.1)
---------------------------------

GM  will be v0.27.0.0 Scheduled: December 2018
RC2 will be v0.27.0.2 Scheduled: November 2018

I have to thank Dan and Luis for their huge contribution to make this release.

I'd like to thank Ben, Freddy, Henri, Michal, Phil and Shridar for their contributions.
If I've forgotten you, please let me know so I can add your name to this acknowledgement.

The headline features are:
1) Many security fixes
2) New CMake build infrastructure
3) New conan support to build dependencies
4) New test harness
5) Support for Adobe XMPsdk 2016
6) Many bug fixes

With Exiv2 v0.27, Team Exiv2 will offer "Long Term Support" and plan "dot" releases with security and critical fixes in 2019 and 2020.

Focus of Exiv2 v0.28
--------------------

Exiv2 moves into the future with Exiv2 v0.28.
The code will be "modernised to C++11 and later".
Users of Elderly compilers such as GCC4 and Visual Studio (<2015) will be supported with 'dot' releases.
To use Exiv2 v0.28, applications must be built with"modern" build systems such as GCC 7+ and Visual Studio 2015 and later.

Known Issues on Redmine for v0.27
---------------------------------

The following issues on Redmine are unresolved and require additional work.
With the exception of #1270, work-arounds have been provided for most of those issues.

http://dev.exiv2.org/projects/exiv2/issues

	0001346	Assertion failure in Exiv2::Internal::CiffComponent::doRead() in libexiv2
	0001345	Assertion Failure (SIGABRT) in Exiv2:RafImage::readMetadata() in Libexiv2
	0001341	Date/Time Original not modified for NEFs
	0001340	Infinite loop bugs in Libexiv2 Exiv2::Image::printIFDStructure()
	0001324	Pentax Makernote written twice
	0001306	Lens TAMRON SP 15-30mm F/2.8 Di VC USD A012 is not properly recognised
	0001304	Lens missing Tamron SP 90mm f/2.8 Di Macro 1:1 VC USD F017
	0001270	Using libexiv2.a/.lib  in multhreaded app segfaults.

Resolved Issues on Redmine for v0.27
------------------------------------

53 issues have been 'closed'
http://dev.exiv2.org/projects/exiv2/issues

Build: (7)
	0001362	Fix Clang build
	0001358	Unsuccessful build (Win7, MSVC2015, amd64_x86)
	0001356	Small patch to configure.py
	0001329	Error during cmake with gcc version parsing
	0001299	exiv2-0.26-trunk.tar.gz changed on download server
	0001270	Using libexiv2.a/.lib  in multhreaded app segfaults.
	0001237	Report CMake/MinGW issues to Kitware

Camera: (1)
	0001352	Support camera Canon M100

Duplicate: (1)
	0001307	exiv2: tiffvisitor.cpp:1299: virtual void Exiv2::Internal::TiffReader::visitDirectory(Exiv2::Internal::TiffDirectory*): Assertion `tc.get()' failed.

Exif: (6)
	0001320	It is a heap-buffer-overflow in Exiv2::Jp2Image::readMetadata (jp2image.cpp:277)
	0001319	It is a heap-buffer-overflow in Exiv2::us2Data (types.cpp:346)
	0001318	 Invalid memory address dereference in Exiv2::StringValueBase::read ( in value.cpp:302)
	0001317	It is a heap-buffer-overflow in Exiv2::s2Data (types.cpp:383)
	0001316	 It is a heap-buffer-overflow in Exiv2::l2Data (types.cpp:398)
	0001315	Invalid memory address dereference in Exiv2::DataValue::read (value.cpp:193)

Insufficient information: (1)
	0001347	Segfault in Digikam when saving/loading certain TIFF images

Image format: (5)
	0001353	Segmentation fault in the software exiv2 when the function Exiv2::tEXtToDataBuf() is finished
	0001346	Assertion failure in Exiv2::Internal::CiffComponent::doRead() in libexiv2
	0001345	Assertion Failure (SIGABRT) in Exiv2:RafImage::readMetadata() in Libexiv2
	0001338	Exiv2 aborted (crashed) with "Error in `exiv2': free(): invalid next size (fast)"
	0001334	Version 0.26 is returns unwanted data when the key value is received from RW2 files (Panasonic DMC-GF6 camera)

Lens: (15)
	0001361	Wrongly identified Canon 2x extender as 1.4x extender with Tamron lens
	0001355	New lens Tamron 100-400mm F/4.5-6.3 Di VC USD
	0001351	Add lens: Sigma 120-300mm f/2.8 G VR
	0001348	patch: add Tamron SP AF 150-600mm F/5-6.3 VC USD G2 (A022) for Nikon
	0001339	Panasonic DMC-TZ70 lens not detected
	0001336	Support for AF-P DX NIKKOR 10-20mm f/4.5-5.6G VR
	0001335	Missing Nikon lense
	0001333	New Lens: Tamron SP 24-70mm F/2.8 Di VC USD G2 (A032)
	0001332	Recognise Tamron 18-200mm F/3.5-6.3 DiII VC B018 on Canon bodies
	0001328	Sigma 18-300mm is wrongly reported as a Canon 75-300mm
	0001327	Two Sigma 24-70mm f/2.8 IF EX DG HSM lens definitions ??
	0001326	Tamron 100-300mm f/5-6.3 MACRO 1:4
	0001323	Add lens "Sigma 18-300mm f/3.5-6.3 DC Macro OS HS" for Canon ID 197
	0001306	the lens TAMRON SP 15-30mm F/2.8 Di VC USD A012 is not properly recognised
	0001304	Lens missing Tamron SP 90mm f/2.8 Di Macro 1:1 VC USD F017

Miscellaneous: (2)
	0001314	it is a stack-overflow vulnerability in Exiv2::Internal::stringFormat[abi:cxx11] ( in image.cpp:975 )
	0001305	Segmentation fault on certain image

Metadata: (7)
	0001341	Date/Time Original not modified for NEFs
	0001340	 Infinite loop bugs in Libexiv2 Exiv2::Image::printIFDStructure()
	0001325	Crash in exiv2  - Nikon1MakerNote::printBarValue (nikonnm.cpp)
	0001322	Exception in PentaxMakerNote::printShutterCount, source: pentaxmn.cpp
	0001321	 Invalid memory address dereference in Exiv2::getULong(types.cpp:246)
	0001301	New lens: Sigma 18-35mm f/1.8 DC HSM Art for Canon
	0001300	New lens: Canon EF-S 35mm f/2.8 Macro IS STM

Makernote: (2)
	0001343	Irix 15mm f/2.4 on Nikon
	0001324	Pentax Makernote written twice

Not-a-bug: (5)
	0001364	New lens Tamron SP 24-70mm F/2.8 Di VC USD G2
	0001350	Add lens: AF-S DX Nikkor 18-140mm
	0001349	Add lens: Sigma 50-100mm F1.8 DC HSM | A
	0001344	Unknown metadata
	0001331	exiv2 fails to delete bulk of metadata in jpeg

Tools: (1)
	0001303	#1 Use 'releases' feature on GitHub

Tiff parser: (3)
	0001360	exiv2 can't read metadata from tiff file (tiff directory length is too large)
	0001359	This does not look like a TIFF image
	0001330	Crash in Exiv2::TiffImage::readMetadata

Website: (2)

src/version.cpp
// ***************************************************************** -*- C++ -*-
/*
 * Copyright (C) 2004-2017 Andreas Huggel <ahuggel@gmx.net>
 *
 * This program is part of the Exiv2 distribution.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, 5th Floor, Boston, MA 02110-1301 USA.
 */
/*
  File:      version.cpp
 */

// *****************************************************************************

#include "config.h"

#ifdef EXV_USE_CURL
#include <curl/curl.h>
#endif

#if defined(__CYGWIN__) || defined(__MINGW__)
#include <windows.h>
# if __LP64__
#  ifdef  _WIN64
#   undef _WIN64
#  endif
#  define _WIN64 1
# endif
#endif

#include "http.hpp"
#include "version.hpp"
#include "makernote_int.hpp"

// Adobe XMP Toolkit
#ifdef EXV_HAVE_XMP_TOOLKIT
#include "xmp_exiv2.hpp"
#endif

// + standard includes
#include <iomanip>
#include <sstream>
#include <string>
#include <vector>
#include <stdio.h>
#include <iostream>

#if defined(EXV_HAVE_REGEX_H)
#include <regex.h>
#endif

// #1147
#ifndef WIN32
#include <unistd.h>
#include <sys/types.h>
#endif

namespace Exiv2 {
    int versionNumber()
    {
        return EXIV2_MAKE_VERSION(EXIV2_MAJOR_VERSION, EXIV2_MINOR_VERSION, EXIV2_PATCH_VERSION);
    }

    std::string versionString()
    {
        std::ostringstream os;
        os << EXIV2_MAJOR_VERSION << '.' << EXIV2_MINOR_VERSION << '.' << EXIV2_PATCH_VERSION;
        return os.str();

    }

    std::string versionNumberHexString()
    {
        std::ostringstream os;
        os << std::hex << std::setw(6) << std::setfill('0') << Exiv2::versionNumber();
        return os.str();
    }

    const char* version()
    {
        return EXV_PACKAGE_VERSION;
    }

    bool testVersion(int major, int minor, int patch)
    {
        return versionNumber() >= EXIV2_MAKE_VERSION(major,minor,patch);
    }
}                                       // namespace Exiv2

#ifndef lengthof
#define lengthof(x) sizeof(x)/sizeof(x[0])
#endif
#ifndef _MAX_PATH
#define _MAX_PATH 512
#endif

// platform specific support for dumpLibraryInfo
#if defined(WIN32)
# include <windows.h>
# include <psapi.h>

// tell MSVC to link psapi.
#ifdef  _MSC_VER
#pragma comment( lib, "psapi" )
#endif

#elif defined(__APPLE__)
# include <mach-o/dyld.h>

#elif defined(__linux__)
# include <unistd.h>
// http://syprog.blogspot.com/2011/12/listing-loaded-shared-objects-in-linux.html
# include <dlfcn.h>
  struct something
  {
    void*  pointers[3];
    struct something* ptr;
  };
  struct lmap
  {
    void*    base_address;   /* Base address of the shared object */
    char*    path;           /* Absolute file name (path) of the shared object */
    void*    not_needed1;    /* Pointer to the dynamic section of the shared object */
    struct lmap *next, *prev;/* chain of loaded objects */
  };
#endif

static bool shouldOutput(const exv_grep_keys_t& greps,const char* key,const std::string& value)
{
    bool bPrint = greps.empty();
    for( exv_grep_keys_t::const_iterator g = greps.begin();
      !bPrint && g != greps.end() ; ++g
    ) {
        std::string Key(key);
#if defined(EXV_HAVE_REGEX_H)
        bPrint = (  0 == regexec( &(*g), key          , 0, NULL, 0)
                 || 0 == regexec( &(*g), value.c_str(), 0, NULL, 0)
                 );
#else
            std::string Pattern(g->pattern_);
            std::string Value(value);
            if ( g->bIgnoreCase_ ) {
                // https://notfaq.wordpress.com/2007/08/04/cc-convert-string-to-upperlower-case/
                std::transform(Pattern.begin(), Pattern.end(),Pattern.begin(), ::tolower);
                std::transform(Key.begin()    , Key.end()    ,Key.begin()    , ::tolower);
                std::transform(Value.begin()  , Value.end()  ,Value.begin()    , ::tolower);
            }
            bPrint = Key.find(Pattern) != std::string::npos || Value.find(Pattern) != std::string::npos;
#endif
    }
    return bPrint;
}

static void output(std::ostream& os,const exv_grep_keys_t& greps,const char* name,const std::string& value)
{
    if ( shouldOutput(greps,name,value) ) os << name << "=" << value << std::endl;
}

static void output(std::ostream& os,const exv_grep_keys_t& greps,const char* name,int value)
{
    std::ostringstream stringStream;
    stringStream << value;
    output(os,greps,name,stringStream.str());
}

void Exiv2::dumpLibraryInfo(std::ostream& os,const exv_grep_keys_t& keys)
{
    Exiv2::StringVector libs; // libs[0] == executable

    int      bits = 8*sizeof(void*);
#if defined(_DEBUG) || defined(DEBUG)
    int debug=1;
#else
    int debug=0;
#endif

#if defined(DLL_EXPORT) || defined(EXV_HAVE_DLL)
    int dll=1;
#else
    int dll=0;
#endif

    const char* compiler =
#if defined(_MSC_VER)
    "MSVC"    ;

#ifndef __VERSION__
    char  version[40];
    sprintf(version,"%d.%02d",(_MSC_VER-600)/100,_MSC_VER%100);

    // add edition in brackets
    // 7.10 = 2003 8.00 = 2005 etc 12.00 = 2013 13.00 = 2015 (yet the installer labels it as 14.0!)
    size_t      edition       = (_MSC_VER-600)/100;
    const char* editions[]    = { "0","1","2","3","4","5","6","2003", "2005", "2008", "2010", "2012","2013","2015","2017"};
    if  ( edition > lengthof(editions) ) edition = 0 ;
    if  ( edition ) sprintf(version+::strlen(version)," (%s/%s)",editions[edition],bits==64?"x64":"x86");
#define __VERSION__ version
#endif

#elif defined(__clang__)
    "Clang"   ;
#elif defined(__GNUG__)
    "G++"     ;
#elif defined(__GNUC__)
    "GCC"     ;
#elif defined(__SUNPRO_CC)
    "CC (oracle)";
#elif defined (__SUNPRO_C)
    "cc (oracle)";
#else
    "unknown" ;
#endif

#if defined(__SUNPRO_CC) || defined (__SUNPRO_C)
#define __oracle__
#endif

#ifndef __VERSION__
#ifdef  __clang__version__
#define __VERSION__ __clang__version__
#else
#define __VERSION__ "unknown"
#endif
#endif

    const char* platform =
#if defined(__CYGWIN__)
    "cygwin";
#elif defined(_MSC_VER)
    "windows";
#elif defined(__APPLE__)
    "apple";
#elif defined(__MINGW64__)
    "mingw64";
#elif defined(__MINGW32__)
    "mingw32";
#elif defined(__linux__)
    "linux";
#else
    "unknown";
#endif

    int have_gmtime_r    =0;
    int have_inttypes    =0;
    int have_libintl     =0;
    int have_lensdata    =0;
    int have_iconv       =0;
    int have_memory      =0;
    int have_lstat       =0;
    int have_stdbool     =0;
    int have_stdint      =0;
    int have_stdlib      =0;
    int have_strlib      =0;
    int have_strerror_r  =0;
    int have_strings_h   =0;
    int have_mmap        =0;
    int have_munmap      =0;
    int have_sys_stat    =0;
    int have_unistd_h    =0;
    int have_sys_mman    =0;
    int have_libz        =0;
    int have_xmptoolkit  =0;
    int adobe_xmpsdk     =0;
    int have_bool        =0;
    int have_strings     =0;
    int have_sys_types   =0;
    int have_unistd      =0;
    int have_unicode_path=0;

    int enable_video     =0;
    int enable_webready  =0;
    int use_curl         =0;
    int use_ssh          =0;

#ifdef EXV_HAVE_GMTIME_R
    have_gmtime_r=1;
#endif

#ifdef EXV_HAVE_INTTYPES_H
    have_inttypes=1;
#endif

#ifdef EXV_HAVE_LIBINTL_H
    have_libintl=1;
#endif

#ifdef EXV_HAVE_LENSDATA
    have_lensdata=1;
#endif

#ifdef EXV_HAVE_ICONV
    have_iconv=1;
#endif

#ifdef EXV_HAVE_LIBINTL_H
    have_libintl=1;
#endif

#ifdef EXV_HAVE_MEMORY_H
    have_memory=1;
#endif

#ifdef EXV_HAVE_LSTAT
    have_lstat=1;
#endif

#ifdef EXV_HAVE_STDBOOL_H
    have_stdbool=1;
#endif

#ifdef EXV_HAVE_STDINT_H
    have_stdint=1;
#endif

#ifdef EXV_HAVE_STDLIB_H
    have_stdlib=1;
#endif

#ifdef EXV_HAVE_STRERROR_R
    have_strerror_r=1;
#endif

#ifdef EXV_HAVE_STRINGS_H
    have_strings=1;
#endif

#ifdef EXV_HAVE_MMAP
    have_mmap=1;
#endif

#ifdef EXV_HAVE_MUNMAP
    have_munmap=1;
#endif

#ifdef EXV_HAVE_SYS_STAT_H
    have_sys_stat=1;
#endif

#ifdef EXV_HAVE_SYS_TYPES_H
    have_sys_types=1;
#endif

#ifdef EXV_HAVE_UNISTD_H
    have_unistd=1;
#endif

#ifdef EXV_HAVE_SYS_MMAN_H
    have_sys_mman=1;
#endif

#ifdef EXV_HAVE_LIBZ
    have_libz=1;
#endif

#ifdef EXV_HAVE_XMP_TOOLKIT
    have_xmptoolkit=1;
#endif

#ifdef EXV_ADOBE_XMPSDK
    adobe_xmpsdk=EXV_ADOBE_XMPSDK;
#endif

#ifdef EXV_HAVE_BOOL
    have_bool=1;
#endif

#ifdef EXV_HAVE_STRINGS
     have_strings=1;
#endif

#ifdef EXV_SYS_TYPES
     have_sys_types=1;
#endif

#ifdef EXV_HAVE_UNISTD
     have_unistd=1;
#endif

#ifdef EXV_UNICODE_PATH
     have_unicode_path=1;
#endif

#ifdef EXV_ENABLE_VIDEO
     enable_video=1;
#endif

#ifdef EXV_ENABLE_WEBREADY
     enable_webready=1;
#endif

#ifdef EXV_USE_CURL
    use_curl=1;
#endif

#ifdef EXV_USE_SSH
     use_ssh=1;
#endif

#if defined(WIN32) || defined(__CYGWIN__) || defined(__MINGW__)
    // enumerate loaded libraries and determine path to executable
    HMODULE handles[200];
    DWORD   cbNeeded;
    if ( EnumProcessModules(GetCurrentProcess(),handles,lengthof(handles),&cbNeeded)) {
        char szFilename[_MAX_PATH];
        for ( DWORD h = 0 ; h < cbNeeded/sizeof(handles[0]) ; h++ ) {
            GetModuleFileNameA(handles[h],szFilename,lengthof(szFilename)) ;
            libs.push_back(szFilename);
        }
    }
#elif defined(__APPLE__)
    // man 3 dyld
    uint32_t count = _dyld_image_count();
    for (uint32_t image = 0 ; image < count ; image++ ) {
        const char* image_path = _dyld_get_image_name(image);
        libs.push_back(image_path);
    }
#elif defined(__linux__)
    // http://stackoverflow.com/questions/606041/how-do-i-get-the-path-of-a-process-in-unix-linux
    char proc[100];
    char path[500];
    sprintf(proc,"/proc/%d/exe", getpid());
    int l = readlink (proc, path,sizeof(path)-1);
    if (l>0) {
        path[l]=0;
        libs.push_back(path);
    } else {
        libs.push_back("unknown");
    }

    // http://syprog.blogspot.com/2011/12/listing-loaded-shared-objects-in-linux.html
    struct lmap*      pl;
    void*             ph = dlopen(NULL, RTLD_NOW);
    struct something* p  = (struct something*) ph;

    p  = p->ptr;
    pl = (struct lmap*)p->ptr;

    while ( pl )
    {
        libs.push_back(pl->path);
        pl = pl->next;
    }
#endif
    output(os,keys,"exiv2",Exiv2::versionString());
    output(os,keys,"platform"       , platform   );
    output(os,keys,"compiler"       , compiler   );
    output(os,keys,"bits"           , bits       );
    output(os,keys,"dll"            , dll        );
    output(os,keys,"debug"          , debug      );
    output(os,keys,"cplusplus"      , __cplusplus);
    output(os,keys,"cplusplus11"    , __cplusplus >= CPLUSPLUS11 );
    output(os,keys,"version"        , __VERSION__);
    output(os,keys,"date"           , __DATE__   );
    output(os,keys,"time"           , __TIME__   );

#ifdef EXV_USE_CURL
    std::string curl_protocols;
    curl_version_info_data* vinfo = curl_version_info(CURLVERSION_NOW);
    for (int i = 0; vinfo->protocols[i]; i++) {
        curl_protocols += vinfo->protocols[i];
        curl_protocols += " " ;
    }
    output(os,keys,"curlprotocols" ,curl_protocols);
#endif

    output(os,keys,"curl"          , use_curl);
    if ( libs.begin() != libs.end() ) {
        output(os,keys,"executable" ,*libs.begin());
        for ( Exiv2::StringVector_i lib = libs.begin()+1 ; lib != libs.end() ; ++lib )
            output(os,keys,"library",*lib);
    }

    output(os,keys,"have_strerror_r"   ,have_strerror_r  );
    output(os,keys,"have_gmtime_r"     ,have_gmtime_r    );
    output(os,keys,"have_inttypes"     ,have_inttypes    );
    output(os,keys,"have_libintl"      ,have_libintl     );
    output(os,keys,"have_lensdata"     ,have_lensdata    );
    output(os,keys,"have_iconv"        ,have_iconv       );
    output(os,keys,"have_memory"       ,have_memory      );
    output(os,keys,"have_lstat"        ,have_lstat       );
    output(os,keys,"have_stdbool"      ,have_stdbool     );
    output(os,keys,"have_stdint"       ,have_stdint      );
    output(os,keys,"have_stdlib"       ,have_stdlib      );
    output(os,keys,"have_strlib"       ,have_strlib      );
    output(os,keys,"have_strerror_r"   ,have_strerror_r  );
    output(os,keys,"have_strings_h"    ,have_strings_h   );
    output(os,keys,"have_mmap"         ,have_mmap        );
    output(os,keys,"have_munmap"       ,have_munmap      );
    output(os,keys,"have_sys_stat"     ,have_sys_stat    );
    output(os,keys,"have_unistd_h"     ,have_unistd_h    );
    output(os,keys,"have_sys_mman"     ,have_sys_mman    );
    output(os,keys,"have_libz"         ,have_libz        );
    output(os,keys,"have_xmptoolkit"   ,have_xmptoolkit  );
    output(os,keys,"adobe_xmpsdk"      ,adobe_xmpsdk     );
    output(os,keys,"have_bool"         ,have_bool        );
    output(os,keys,"have_strings"      ,have_strings     );
    output(os,keys,"have_sys_types"    ,have_sys_types   );
    output(os,keys,"have_unistd"       ,have_unistd      );
    output(os,keys,"have_unicode_path" ,have_unicode_path);
    output(os,keys,"enable_video"      ,enable_video     );
    output(os,keys,"enable_webready"   ,enable_webready  );
    output(os,keys,"use_curl"          ,use_curl         );
    output(os,keys,"use_ssh"           ,use_ssh          );

    output(os,keys,"config_path"       ,Exiv2::Internal::getExiv2ConfigPath());

// #1147
#ifndef WIN32
    uid_t uid  = getuid()  ; output(os,keys,"uid" ,  uid  );
    uid_t euid = geteuid() ; output(os,keys,"euid", euid  );
    uid_t gid  = getgid()  ; output(os,keys,"gid" ,  gid  );
#endif

#ifdef EXV_HAVE_XMP_TOOLKIT
    const char* name = "xmlns";

    Exiv2::Dictionary ns;
    Exiv2::XmpProperties::registeredNamespaces(ns);
    for ( Exiv2::Dictionary_i it = ns.begin(); it != ns.end() ; ++it ) {
        std::string xmlns = (*it).first;
        std::string uri   = (*it).second;
        output(os,keys,name,xmlns+":"+uri);
    }
#endif

#if defined(__linux__)
    dlclose(ph);
    ph=NULL;
#endif

}
